parser/proto.y

```diff
@@ -327,269 +334,231 @@ stringLit : _STRING_LIT {
 	}
 
 msgElementTypeIdent : msgElementIdent {
 		$$ = $1.toIdentValueNode(nil)
 	}
-	| '.' ident {
+	| '.' qualifiedIdentifier {
 		$$ = $2.toIdentValueNode($1)
 	}
 
 extElementTypeIdent : extElementIdent {
 		$$ = $1.toIdentValueNode(nil)
 	}
-	| '.' ident {
+	| '.' qualifiedIdentifier {
 		$$ = $2.toIdentValueNode($1)
 	}
 
 oneofElementTypeIdent : oneofElementIdent {
 		$$ = $1.toIdentValueNode(nil)
 	}
-	| '.' ident {
+	| '.' qualifiedIdentifier {
 		$$ = $2.toIdentValueNode($1)
 	}
 
-msgField : _REQUIRED typeIdent name '=' _INT_LIT ';' {
-		$$ = ast.NewFieldNode($1.ToKeyword(), $2, $3, $4, $5, nil, $6)
+fieldDecl : fieldCardinality typeName identifier '=' _INT_LIT ';' {
+		$$ = ast.NewFieldNode($1, $2, $3, $4, $5, nil, $6)
 	}
-	| _OPTIONAL typeIdent name '=' _INT_LIT ';' {
-		$$ = ast.NewFieldNode($1.ToKeyword(), $2, $3, $4, $5, nil, $6)
+	| fieldCardinality typeName identifier '=' _INT_LIT compactOptions ';' {
+		$$ = ast.NewFieldNode($1, $2, $3, $4, $5, $6, $7)
 	}
-	| _REPEATED typeIdent name '=' _INT_LIT ';' {
-		$$ = ast.NewFieldNode($1.ToKeyword(), $2, $3, $4, $5, nil, $6)
-	}
-	| _REQUIRED typeIdent name '=' _INT_LIT compactOptions ';' {
-		$$ = ast.NewFieldNode($1.ToKeyword(), $2, $3, $4, $5, $6, $7)
-	}
-	| _OPTIONAL typeIdent name '=' _INT_LIT compactOptions ';' {
-		$$ = ast.NewFieldNode($1.ToKeyword(), $2, $3, $4, $5, $6, $7)
-	}
-	| _REPEATED typeIdent name '=' _INT_LIT compactOptions ';' {
-		$$ = ast.NewFieldNode($1.ToKeyword(), $2, $3, $4, $5, $6, $7)
-	}
-	| msgElementTypeIdent name '=' _INT_LIT ';' {
+	| msgElementTypeIdent identifier '=' _INT_LIT ';' {
 		$$ = ast.NewFieldNode(nil, $1, $2, $3, $4, nil, $5)
 	}
-	| msgElementTypeIdent name '=' _INT_LIT compactOptions ';' {
+	| msgElementTypeIdent identifier '=' _INT_LIT compactOptions ';' {
 		$$ = ast.NewFieldNode(nil, $1, $2, $3, $4, $5, $6)
 	}
 
-extField : _REQUIRED typeIdent name '=' _INT_LIT ';' {
-		$$ = ast.NewFieldNode($1.ToKeyword(), $2, $3, $4, $5, nil, $6)
-	}
-	| _OPTIONAL typeIdent name '=' _INT_LIT ';' {
-		$$ = ast.NewFieldNode($1.ToKeyword(), $2, $3, $4, $5, nil, $6)
-	}
-	| _REPEATED typeIdent name '=' _INT_LIT ';' {
-		$$ = ast.NewFieldNode($1.ToKeyword(), $2, $3, $4, $5, nil, $6)
-	}
-	| _REQUIRED typeIdent name '=' _INT_LIT compactOptions ';' {
-		$$ = ast.NewFieldNode($1.ToKeyword(), $2, $3, $4, $5, $6, $7)
-	}
-	| _OPTIONAL typeIdent name '=' _INT_LIT compactOptions ';' {
-		$$ = ast.NewFieldNode($1.ToKeyword(), $2, $3, $4, $5, $6, $7)
+fieldCardinality : _REQUIRED {
+		$$ = $1.ToKeyword()
 	}
-	| _REPEATED typeIdent name '=' _INT_LIT compactOptions ';' {
-		$$ = ast.NewFieldNode($1.ToKeyword(), $2, $3, $4, $5, $6, $7)
+	| _OPTIONAL {
+		$$ = $1.ToKeyword()
 	}
-	| extElementTypeIdent name '=' _INT_LIT ';' {
-		$$ = ast.NewFieldNode(nil, $1, $2, $3, $4, nil, $5)
-	}
-	| extElementTypeIdent name '=' _INT_LIT compactOptions ';' {
-		$$ = ast.NewFieldNode(nil, $1, $2, $3, $4, $5, $6)
+	| _REPEATED {
+		$$ = $1.ToKeyword()
 	}
```

Comment:
If you change the type of `fieldCardinality` to be `%type<id>`, you can remove `fldCard` from the union type and then also simplify this (moving calls to `ToKeyword()` to the productions above):
```yacc
fieldCardinality : _REQUIRED
	| _OPTIONAL
	| _REPEATED
```

Reply:
I changed `fieldCardinality` to be `%type<id>` and simply the production.